leader-election-impl/src/main/java/com/palantir/paxos/SingleLeaderPinger.java

```diff
@@ -58,44 +59,49 @@ public SingleLeaderPinger(
 
     @Override
     public LeaderPingResult pingLeaderWithUuid(UUID uuid) {
-        Optional<PingableLeader> suspectedLeader = getSuspectedLeader(uuid);
+        Optional<LeaderPingerContext<PingableLeader>> suspectedLeader = getSuspectedLeader(uuid);
         if (!suspectedLeader.isPresent()) {
             return LeaderPingResults.pingReturnedFalse();
         }
 
-        PingableLeader leader = suspectedLeader.get();
+        LeaderPingerContext<PingableLeader> leader = suspectedLeader.get();
 
-        MultiplexingCompletionService<PingableLeader, Boolean> multiplexingCompletionService
+        MultiplexingCompletionService<LeaderPingerContext<PingableLeader>, Boolean> multiplexingCompletionService
                 = MultiplexingCompletionService.create(leaderPingExecutors);
 
-        multiplexingCompletionService.submit(leader, leader::ping);
+        multiplexingCompletionService.submit(leader, () -> leader.pinger().ping());
 
         try {
-            Future<Map.Entry<PingableLeader, Boolean>> pingFuture = multiplexingCompletionService.poll(
+            Future<Map.Entry<LeaderPingerContext<PingableLeader>, Boolean>> pingFuture = multiplexingCompletionService.poll(
                     leaderPingResponseWait.toMillis(),
                     TimeUnit.MILLISECONDS);
-            return getAndRecordLeaderPingResult(pingFuture);
+            return getLeaderPingResult(uuid, pingFuture);
         } catch (InterruptedException ex) {
             Thread.currentThread().interrupt();
             return LeaderPingResults.pingCallFailure(ex);
         }
     }
 
-    private static LeaderPingResult getAndRecordLeaderPingResult(
-            @Nullable Future<Map.Entry<PingableLeader, Boolean>> pingFuture) throws InterruptedException {
+    private static LeaderPingResult getLeaderPingResult(
```

Comment:
Would it work if instead we just had a method to return HostAndPort for a UUID? then we wouldn't need to change the LeaderPingResult at all, and I feel the diff would be much smaller.

Reply:
Not necessarily, to do that, we'd need `pinghLeaderWithUuid` to record the same amount of inforamtion it's doing now, and save it locally, and then expose that it up via a method you method. Similarly for `AutobatchingPingableLeaderFactory`. In a sense, it'd look like ping, except that it would return host and port, this feels pretty weird imo.

Throughout, you need to keep the association, so you can't really get rid of it.